prow/tide/tide.go

```diff
@@ -0,0 +1,289 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package tide contains a controller for managing a tide pool of PRs.
+package tide
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/shurcooL/githubql"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/kube"
+)
+
+type githubClient interface {
+	GetRef(string, string, string) (string, error)
+	Query(context.Context, interface{}, map[string]interface{}) error
+}
+
+// Controller knows how to sync PRs and PJs.
+type Controller struct {
+	log *logrus.Entry
+	ca  *config.Agent
+	ghc githubClient
+	kc  *kube.Client
+}
+
+// NewController makes a Controller out of the given clients.
+func NewController(l *logrus.Entry, ghc *github.Client, kc *kube.Client, ca *config.Agent) *Controller {
+	return &Controller{
+		log: l,
+		ghc: ghc,
+		kc:  kc,
+		ca:  ca,
+	}
+}
+
+// Sync runs one sync iteration.
+func (c *Controller) Sync() error {
+	ctx := context.Background()
+	c.log.Info("Building tide pool.")
+	var pool []pullRequest
+	for _, q := range c.ca.Config().Tide.Queries {
+		prs, err := c.search(ctx, q)
+		if err != nil {
+			return err
+		}
+		pool = append(pool, prs...)
+	}
+	var pjs []kube.ProwJob
+	var err error
+	if len(pool) > 0 {
+		c.log.Info("Listing ProwJobs.")
+		pjs, err = c.kc.ListProwJobs(nil)
+		if err != nil {
+			return err
+		}
+	}
+	sps, err := c.dividePool(pool, pjs)
+	if err != nil {
+		return err
+	}
+	for _, sp := range sps {
+		if err := c.syncSubpool(ctx, sp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type simpleState string
+
+const (
+	noneState    simpleState = "none"
+	pendingState simpleState = "pending"
+	successState simpleState = "success"
+)
+
+func toSimpleState(s kube.ProwJobState) simpleState {
+	if s == kube.TriggeredState || s == kube.PendingState {
+		return pendingState
+	} else if s == kube.SuccessState {
+		return successState
+	}
+	return noneState
+}
+
+func pickSmallestNumber(prs []pullRequest) pullRequest {
+	smallestNumber := -1
+	var smallestPR pullRequest
+	for _, pr := range prs {
+		if smallestNumber < 0 || int(pr.Number) < smallestNumber {
+			smallestNumber = int(pr.Number)
+			smallestPR = pr
+		}
+	}
+	return smallestPR
+}
+
+// TODO(spxtr): Batch merges. These complicate the logic further. One
+// particularly nasty complication is that we want to merge PRs even if the
+// serial test for a PR failed but the batch succeeded. This means that the PR
```

Comment:
This is not true today, right? While I like the end goal, I also think we should try to incentivize devs to fix flakes. :)